libs/muddle/include/muddle/muddle_interface.hpp

```diff
@@ -0,0 +1,243 @@
+  class Identity;
+}
+
+namespace network {
+  class NetworkManager;
+}
+
+namespace muddle {
+
+class MuddleEndpoint;
+class NetworkId;
+
+class MuddleInterface
+{
+public:
+  enum class Confidence
+  {
+    DEFAULT,
+    WHITELIST,
+    BLACKLIST
+  };
+
+  using Peers         = std::unordered_set<std::string>;
+  using Ports         = std::vector<uint16_t>;
+  using Addresses     = std::unordered_set<Address>;
+  using ConfidenceMap = std::unordered_map<Address, Confidence>;
+  using AddressHints  = std::unordered_map<Address, network::Uri>;
+
+  // Construction / Destruction
+  MuddleInterface() = default;
+  virtual ~MuddleInterface() = default;
+
+  /// @name Muddle Setup
+  /// @{
+
+  /**
+   * Start the muddle instance connecting to the initial set of peers and listing on the specified
+   * set of ports
+   *
+   * @param peers The initial set of peers that muddle should connect to
+   * @param ports The set of ports to listen on. Zero signals a random port
+   * @return true if successful, otherwise false
+   */
+  virtual bool Start(Peers const &peers, Ports const &ports) = 0;
+
+  /**
+   * Stop the muddle instance, this will cause all the connected to close
+   */
+  virtual void Stop() = 0;
+
+  /**
+   * Get the endpoint interface for this muddle instance/
+   *
+   * @return The endpoint pointer
+   */
+  virtual MuddleEndpoint &GetEndpoint() = 0;
+
+  /// @}
+
+  /// @name Muddle Status
+  /// @{
+
+  /**
+   * Get the associated network for this muddle instance
+   *
+   * @return The current network
+   */
+  virtual NetworkId const &GetNetwork() const = 0;
+
+  /**
+   * Get the address of this muddle node
+   *
+   * @return The address of the node
+   */
+  virtual Address const &GetAddress() const = 0;
+
+  /**
+   * Get the set of ports that the server is currently listening on
+   *
+   * @return The set of server ports
+   */
+  virtual Ports GetListeningPorts() const = 0;
+
+  /**
+   * Get the set of addresses to whom this node is directly connected to
+   *
+   * @return The set of addresses
+   */
+  virtual Addresses GetDirectlyConnectedPeers() const = 0;
+
+  /**
+   * Get the set of addresses of peers that are connected directly to this node
+   *
+   * @return The set of addresses
+   */
+  virtual Addresses GetIncomingConnectedPeers() const = 0;
+
+  /**
+   * Get the set of addresses of peers that we are directly connected to
+   *
+   * @return The set of peers
+   */
+  virtual Addresses GetOutgoingConnectedPeers() const = 0;
+
+
+  /**
+   * Get the number of peers that are directly connected to this node
+   *
+   * @return The number of directly connected peers
+   */
+  virtual std::size_t GetNumDirectlyConnectedPeers() const = 0;
+
+  /**
+   * Determine if we are directly connected to the specified address
+   *
+   * @param address The address to check
+   * @return true if directly connected, otherwise false
+   */
+  virtual bool IsDirectlyConnected(Address const &address) const= 0;
```

Comment:
Let's think about how hard it would be to make a `IsReachable` function as well? In many cases, you care more about this than whether you can get a direct connection.

Reply:
It would be reasonably easy but might be quite unreliable